caom2pipe/execute_composable.py

```diff
@@ -1382,14 +1414,25 @@ def choose(self, storage_name):
         executors = []
         if mc.TaskType.SCRAPE in self.task_types:
             cred_param = None
-            cadc_data_client = None
+            cadc_client = None
             caom_repo_client = None
         else:
             subject, cred_param = self._define_subject()
-            cadc_data_client = CadcDataClient(subject)
             caom_repo_client = CAOM2RepoClient(
                 subject, self.config.logging_level,
                 self.config.resource_id)
+            if self.config.features.supports_latest_client:
+                self._logger.warning('Using vos.Client for storage.')
+                cadc_client = Client(
+                    vospace_certfile=self._define_client_credentials())
+            else:
+                self._logger.warning(
+                    'Using cadcdata.CadcDataClient for storage.')
+                cadc_client = CadcDataClient(subject)
+            for entry in [self._modify_transfer, self._store_transfer]:
```

Comment:
I added this comment to the code:
```
            # Provide access to a single client instance for transactions with
            # CADC storage services. If the Transfer specialization doesn't
            # have a cadc_client attribute, it's an external data source that
            # manages its own connection.
```

Does that help? Or should this be handled in a different way?

Reply:
The logic here is still not clear to me. Is the purpose to override the clients of the transfers instances passed to this object? Is that OK to do? Are those transfers not used elsewhere outside this class?

Reply:
Add more comments:

            # The cadc_client instances are provided externally to the the 
            # classes that use them, so that it’s always the same instance 
            # of the client, reducing the number of connections in use.

Reply:
If they are provided from outside why are they overriden?
I'm not sure cadc_client maintains the same connection (that depends on the `request` library) but it might have some overhead on instantiation (discovering the capabilities etc.). If you want to reuse them why not setting `cadc_client` with one of the external instances if they exist. Also, how do we know whether the external transfers have a data or a vospace client when that gets overriden. I'm pretty sure I'm missing something but it's not clear.

Reply:
For re-use: `cadc_client` is only set in this location, so there are no external instances besides the one that is used to set the property here.

For external transfers: these classes (HttpTransfer, FtpTransfer) do not have a cadc_client member. The Transfer classes exist only to retrieve data from somewhere.

The `cadc_client` data member is set at this time in the execution, because this is when the `Config` class with its `TaskType` selections and `FeatureFlag` , the `composable.py` from `gem2caom2`, and the `CaomExecutor` extensions are accessible at time/location to make the correct decisions about which clients are required for what purpose.

I'd rather have this back and forth so that the idea is clear, or I figure out a better way to do things.